libazureinit/src/provision/user.rs

```diff
@@ -96,28 +100,50 @@ impl Provisioner {
     }
 }
 
+#[instrument(skip_all)]
+fn user_exists(username: &str) -> Result<bool, Error> {
+    let status = Command::new("getent")
+        .arg("passwd")
+        .arg(username)
+        .status()?;
+
+    Ok(status.success())
+}
+
 #[instrument(skip_all)]
 fn useradd(user: &User) -> Result<(), Error> {
-    let path_useradd = env!("PATH_USERADD");
-    let home_path = format!("/home/{}", user.name);
-
-    let status = Command::new(path_useradd)
-                    .arg(&user.name)
-                    .arg("--comment")
-                    .arg(
-                      "Provisioning agent created this user based on username provided in IMDS",
-                    )
-                    .arg("--groups")
-                    .arg(user.groups.join(","))
-                    .arg("-d")
-                    .arg(home_path)
-                    .arg("-m")
-                    .status()?;
-    if !status.success() {
-        return Err(Error::SubprocessFailed {
-            command: path_useradd.to_string(),
-            status,
-        });
+    const MAX_RETRIES: u8 = 5;
+
+    if user_exists(&user.name)? {
+        return Ok(());
+    }
+
+    for attempt in 1..=MAX_RETRIES {
+        let path_useradd = env!("PATH_USERADD");
+        let home_path = format!("/home/{}", user.name);
+
+        let status = Command::new(path_useradd)
+                        .arg(&user.name)
+                        .arg("--comment")
+                        .arg(
+                          "Provisioning agent created this user based on username provided in IMDS",
+                        )
+                        .arg("--groups")
+                        .arg(user.groups.join(","))
+                        .arg("-d")
+                        .arg(home_path)
+                        .arg("-m")
+                        .status()?;
+        if status.success() {
+            return Ok(());
+        } else if attempt < MAX_RETRIES {
+            sleep(Duration::from_secs(30));
```

Comment:
Based on the discussion above this is going to go away, but I think it's worth noting that using `std::thread::sleep` could have some unexpected side effects since we might be executing in an asynchronous context here. 

In this case, the `azure-init` binary calls this (indirectly, via `libazure::Provision::provision`) from a thread in the tokio runtime. If we use `std::thread::sleep` we'll put that thread to sleep and tokio will not be able to complete work on that thread while it's asleep. If the user has other asynchronous tasks running, they will see everything freeze for 30 seconds before continuing. `tokio` ships its own [sleep](https://docs.rs/tokio/latest/tokio/time/fn.sleep.html) which will do what you want but allow tokio to keep working on other tasks. To use that, though, we'd need to mark the function as asynchronous.

In general (not just in Rust), we want to be careful about sleeping in library functions. That's not to say one should never sleep in a library: having retries for fallible things is nice for users, but you'll want to note in the API documentation that it will sleep between retries.

Reply:
Super good to know -- forgot about the asynchronous context so very very good to know. Thank you for this!! I will file this information for the future.